stan/math/prim/err/check_flag_sundials.hpp

```diff
@@ -1,12 +1,178 @@
 #ifndef STAN_MATH_PRIM_ERR_CHECK_FLAG_SUNDIALS_HPP
 #define STAN_MATH_PRIM_ERR_CHECK_FLAG_SUNDIALS_HPP
 
+#include <kinsol/kinsol.h>
+#include <cvodes/cvodes.h>
 #include <stan/math/prim/meta.hpp>
-#include <stan/math/prim/err/throw_domain_error.hpp>
+#include <stan/math/prim/err/domain_error.hpp>
 
 namespace stan {
 namespace math {
 
+#define CHECK_CVODES_CALL(call) cvodes_check(call, #call)
+#define CHECK_KINSOL_CALL(call) kinsol_check(call, #call)
+
+/**
+ * Map cvodes error flag to acutally error msg. The most frequent
+ * errors are put at the top. An alternative would be to use std::map
+ * but in our case the difference would be negligible. Note that we
+ * don't use CVGetReturnFlagName function to retrieve the constant
+ * because sanitizer indicates it contains mem leak.
+ *
+ * @param flag
+ *
+ * @return error msg string constant and actuall informative msg
+ */
+inline std::array<std::string, 2> cvodes_flag_msg(int flag) {
+  std::array<std::string, 2> msg;
+  switch (flag) {
+    case -1:
+      msg = {"CV_TOO_MUCH_WORK",
+             "The solver took mxstep internal steps but could not reach tout"};
+      break;  // NOLINT
+    case -2:
+      msg = {"CV_TOO_MUCH_ACC",
+             "The solver could not satisfy the accuracy demanded by the user "
+             "for some internal step"};
+      break;  // NOLINT
+    case -3:
+      msg = {"CV_ERR_FAILURE",
+             "Error test failures occurred too many times during one internal "
+             "time step or minimum step size was reached"};
+      break;  // NOLINT
+    case -4:
+      msg = {"CV_CONV_FAILURE",
+             "Convergence test failures occurred too many times during one "
+             "internal time step or minimum step size was reached"};
+      break;  // NOLINT
+    case -8:
+      msg = {"CV_RHSFUNC_FAIL",
+             "The right-hand side function failed in an unrecoverable manner"};
+      break;  // NOLINT
+    case -9:
+      msg = {"CV_FIRST_RHSFUNC_ERR",
+             "The right-hand side function failed at the first call"};
+      break;  // NOLINT
+    case -10:
+      msg = {"CV_REPTD_RHSFUNC_ERR",
+             "The right-hand side function had repetead recoverable errors"};
+      break;  // NOLINT
+    case -11:
+      msg = {"CV_UNREC_RHSFUNC_ERR",
+             "The right-hand side function had a recoverable error, but no "
+             "recovery is possible"};
+      break;  // NOLINT
+    case -27:
+      msg = {"CV_TOO_CLOSE",
+             "The output and initial times are too close to each other"};
+      break;  // NOLINT
+    default:
```

Comment:
I am not sure I completely understand why there is another switch inside the default? Shouldn't this just be a one-level switch-case statement?

Reply:
I put most common cases at the top. The rest are in a 2nd level of switch. IMO it helps to read, may also help compiler optimize but in our case it doesn't really matter.

Reply:
Makes sense.